### PR TITLE
fix(form): save errors should be more recognizable for the user

### LINKED ISSUE #79: ps-form: Save errors should be more recognizable for the user
If the save callback of the form runs into an error, a card with the error information is displayed at the bottom of the form. However, if the form is longer than the current view, the user cannot see the error immediately. For him it looks like everything went fine.

A possible solution would be to show an error icon next to the save button. When the user clicks on it, the view scrolls to the error card.

For implementation we could use the [IntersectionObserver](https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver) and the [ScrollIntoView](https://developer.mozilla.org/de/docs/Web/API/Element/scrollIntoView) functionality.

### PR DESCRIPTION
**Pull request by 6rayWolf**
fix #79

### CODE REVIEW

File: projects/components/form/src/form.component.ts [hunk 1]
```diff
@@ -390,4 +407,32 @@ export class PsFormComponent implements OnChanges, OnInit, OnDestroy {
...
+        if (!this._errorCardObserver) {
+          const options = {
+            root: null as any, // relative to document viewport
+            rootMargin: '-100px', // margin around root. Values are similar to css property. Unitless values not allowed
+            threshold: 1.0, // visible amount of item shown in relation to root
+          };
+
+          this._errorCardObserver = new IntersectionObserver((changes, _) => {
+            this._isErrorCardInView = changes[0].intersectionRatio > 0;
+            console.log(changes[0].intersectionRatio);
```
Comment (saithis) [inline projects/components/form/src/form.component.ts:423]: This should be removed ;)

File: projects/components/form/src/form.component.ts [hunk 2]
```diff
@@ -390,4 +407,32 @@ export class PsFormComponent implements OnChanges, OnInit, OnDestroy {
...
+          }, options);
+
+          const errorCard = document.querySelector('.ps-form__error-container');
+          this._errorCardObserver.observe(errorCard);
+        }
+      } else {
+        this._isErrorCardInView = isInViewport(document.querySelector('.ps-form__error-container'));
```
[the remark this answers is not in the document]
Comment (saithis) [reply projects/components/form/src/form.component.ts:430]: Also this doesn't work in all cases: Wenn the error card is in initially view, then I scroll up so it isn't anymore, then the button doesn't work to scroll me back down anymore.

File: projects/components/form/src/form.component.ts [hunk 3]
```diff
@@ -390,4 +407,32 @@ export class PsFormComponent implements OnChanges, OnInit, OnDestroy {
     }
     this._savebar.cd.markForCheck();
   }
+
+  private setupIntersectionObserver(ex: IPsFormException) {
+    if (ex) {
+      if ('IntersectionObserver' in window) {
```
Comment (saithis) [inline projects/components/form/src/form.component.ts:413]: All browsers we support also support IntersectionObserver. Only IE doesn't, but our components don't fully work with IE anyway.

File: projects/components/form/src/form.component.ts [hunk 4]
```diff
@@ -390,4 +407,32 @@ export class PsFormComponent implements OnChanges, OnInit, OnDestroy {
...
+            rootMargin: '-100px', // margin around root. Values are similar to css property. Unitless values not allowed
+            threshold: 1.0, // visible amount of item shown in relation to root
+          };
+
+          this._errorCardObserver = new IntersectionObserver((changes, _) => {
+            this._isErrorCardInView = changes[0].intersectionRatio > 0;
+            console.log(changes[0].intersectionRatio);
+          }, options);
+
+          const errorCard = document.querySelector('.ps-form__error-container');
```
Comment (saithis) [inline projects/components/form/src/form.component.ts:426]: Please use @ViewChild

File: projects/components/form/src/form.component.html [hunk 5]
```diff
@@ -37,13 +37,18 @@
         <ng-container *ngTemplateOutlet="content"></ng-container>
       </ps-block-ui>
 
-      <mat-card *ngIf="exception" class="ps-form__error-container" [class.ps-form__error-container--center]="exception.alignCenter">
-        <mat-icon *ngIf="exception.icon" class="ps-form__error-icon">{{ exception.icon }}</mat-icon>
-        <span>{{ exceptionMessage }}</span>
-      </mat-card>
+      <div #errorCardWrapper class="ps-form__error-wrapper">
```
Comment (saithis) [inline projects/components/form/src/form.component.html:40]: the css class is not neccessary

File: projects/components/form/src/form.component.spec.ts [hunk 6]
```diff
@@ -248,6 +251,68 @@ describe('PsFormComponent', () => {
...
+      dataSource.cdTrigger$.next();
+      fixture.detectChanges();
+
+      expect(getErrorContainer(fixture)).not.toBe(null);
+      expect(getErrorNotifier(fixture)).not.toBe(null);
+
+      expect(isInViewport(component.formComponent.errorCardWrapper.nativeElement)).toBe(false);
+
+      const errorNotifier = getErrorNotifier(fixture);
+      const errorWrapper = getErrorWrapper(fixture);
```
Comment (saithis) [inline projects/components/form/src/form.component.spec.ts:302]: you could also get this via component.formComponent.errorCardWrapper

File: projects/components/form/src/form.component.ts [hunk 7]
```diff
@@ -76,7 +77,14 @@ export class PsFormComponent implements OnChanges, OnInit, OnDestroy {
     this._dataSource = value;
 
     if (this._dataSource) {
+      this.createErrorCardObserver();
       this._dataSourceSub = this._dataSource.connect().subscribe(() => {
+        if (!this._dataSource.exception) {
+          this._errorCardObserver.unobserve(this.errorCardWrapper.nativeElement);
+        } else if (this._dataSource.exception !== this._lastException) {
```
Comment (saithis) [inline projects/components/form/src/form.component.ts:84]: If the exception changes from one to another, then you observe the wrapper twice.
You should rather check if you already observe the wrapper or not. (like what you did in the last commit)

File: projects/components/form/src/form.component.ts [hunk 8]
```diff
@@ -390,4 +413,18 @@ export class PsFormComponent implements OnChanges, OnInit, OnDestroy {
...
     this._savebar.cd.markForCheck();
   }
+
+  private createErrorCardObserver() {
+    if (!this._errorCardObserver) {
+      const options = {
+        root: null as any, // relative to document viewport
+        rootMargin: '-100px', // margin around root. Values are similar to css property. Unitless values not allowed
+        threshold: 1.0, // visible amount of item shown in relation to root
+      };
```
Comment (saithis) [inline projects/components/form/src/form.component.ts:423]: I think we should add a delay here to the options

File: projects/components/form/src/form.component.html [hunk 9]
```diff
@@ -37,13 +37,18 @@
...
+        <mat-card *ngIf="exception" class="ps-form__error-container" [class.ps-form__error-container--center]="exception.alignCenter">
+          <mat-icon *ngIf="exception.icon" class="ps-form__error-icon">{{ exception.icon }}</mat-icon>
+          <span>{{ exceptionMessage }}</span>
+        </mat-card>
+      </div>
     </div>
     <div *psSavebarRightContent class="ps-form__buttons">
       <ng-content select="[psFormSavebarButtons]"></ng-content>
+      <mat-icon *ngIf="exception" color="warn" class="ps-form__error-notfier-button" (click)="onErrorNotifierClick()"
```
Comment (saithis) [inline projects/components/form/src/form.component.html:49]: maybe bind this icon to !_isErrorCardInView ? If the error is visible, then we don't realy need this icon.

File: projects/components/form/src/form.component.ts (hunk 7 above)
Comment (6rayWolf) [reply projects/components/form/src/form.component.ts:84]: The IntersectionObserver seems to track the element instances which it observes. That means even if you try to observe the same element multiple times, the observer triggers only once per element instance.

File: projects/components/form/src/form.component.ts (hunk 8 above)
Comment (6rayWolf) [reply projects/components/form/src/form.component.ts:423]: It seems like the typescript implementation of the IntersectionObserver doesn't have such options yet. [see here](https://microsoft.github.io/PowerBI-JavaScript/interfaces/_node_modules_typedoc_node_modules_typescript_lib_lib_dom_d_.intersectionobserver.html)

File: projects/components/form/src/form.component.spec.ts [hunk 10]
```diff
@@ -312,6 +311,93 @@ describe('PsFormComponent', () => {
...
+
+      opts.errorInView$.subscribe(value => errorInViewCheck(value));
+
+      ds.exception = { errorObject: new Error('asdf') };
+      ds.cdTrigger$.next();
+      fixture.detectChanges();
+
+      expect(getErrorContainer(fixture)).not.toBe(null);
+
+      errorInViewCheck = (value: boolean) => expect(value).toBe(true);
```
Comment (saithis) [inline projects/components/form/src/form.component.spec.ts:365]: If errorInView$ doesnt emit, then the assert wont be executed. You could make an array and push the emitted values into the array. Then check the content of the array outside of the subscription.

File: projects/components/form/src/form.component.spec.ts [hunk 11]
```diff
@@ -312,6 +311,93 @@ describe('PsFormComponent', () => {
...
+      let opts: IPsFormDataSourceConnectOptions;
+      spyOn(ds, 'connect').and.callFake((options: IPsFormDataSourceConnectOptions) => {
+        opts = options;
+        return ds.cdTrigger$;
+      });
+      fixture.detectChanges();
+      expect(ds.connect).toHaveBeenCalledTimes(1);
+      expect(opts).toBeDefined();
+
+      let errorInViewCheck = (value: boolean) => expect(value).toBe(false);
```
Comment (saithis) [inline projects/components/form/src/form.component.spec.ts:385]: The same here.

File: projects/components/form/src/form.component.spec.ts [hunk 12]
```diff
@@ -312,6 +311,93 @@ describe('PsFormComponent', () => {
...
+      ds.cdTrigger$.next();
+      fixture.detectChanges();
+
+      expect(getErrorContainer(fixture)).not.toBe(null);
+
+      errorInViewCheck = (value: boolean) => expect(value).toBe(true);
+      opts.scrollToError();
+    });
+
+    it('should update call errorCardWrappers scrollIntoView only once', async () => {
```
Comment (saithis) [inline projects/components/form/src/form.component.spec.ts:369]: -update

**Pull request merged**